### ISSUE TITLE
Naming conventions for extension structs and implementation-specific extensions

### ISSUE THREAD

**Issue by kainino0x** [webgpu-native/webgpu-headers#212]
We should define some naming conventions for both standardized extension structs and implementation-specific extensions. Right now we have the following (inconsistent) names:

`WGPUPrimitiveState` extensions:
`WGPUPrimitiveDepthClipControl`

`WGPURenderPassDescriptor` extensions:
`WGPURenderPassDescriptorMaxDrawCount`

`WGPUShaderModuleDescriptor` extensions:
`WGPUShaderModuleSPIRVDescriptor`
`WGPUShaderModuleWGSLDescriptor`

`WGPUSurfaceDescriptor` extensions:
`WGPUSurfaceDescriptorFromAndroidNativeWindow`
`WGPUSurfaceDescriptorFromCanvasHTMLSelector`
`WGPUSurfaceDescriptorFromMetalLayer`
`WGPUSurfaceDescriptorFromWaylandSurface`
`WGPUSurfaceDescriptorFromWindowsHWND`
`WGPUSurfaceDescriptorFromXcbWindow`
`WGPUSurfaceDescriptorFromXlibWindow`

Additionally for Dawn we might want things like `WGPUSomeDescriptorNewFeatureDAWN`? (though idk what the naming would be for wgpu-native)

**Comment by Kangz** [comment]
For ergonomics it seems that typing `Descriptor` all the time is annoying. So I'd favor a rule like:

 - Take the name of the chain root, remove `Descriptor`, `State` or other common words
 - Optionally add a name to scope to a project / vendor `Dawn`, `Wgpu`, ...
 - Add the extra words.

So the list would become:

`WGPUPrimitiveState` extensions:
`WGPUPrimitiveDepthClipControl`

`WGPURenderPassDescriptor` extensions:
`WGPURenderPassMaxDrawCount`

`WGPUShaderModuleDescriptor` extensions:
`WGPUShaderModuleSPIRV`
`WGPUShaderModuleWGSL`

`WGPUSurfaceDescriptor` extensions:
`WGPUSurfaceFromAndroidNativeWindow`
`WGPUSurfaceFromCanvasHTMLSelector`
`WGPUSurfaceFromMetalLayer`
`WGPUSurfaceFromWaylandSurface`
`WGPUSurfaceFromWindowsHWND`
`WGPUSurfaceFromXcbWindow`
`WGPUSurfaceFromXlibWindow`

WDYT?

**Comment by kainino0x** [comment]
webgpu.h meeting resolution:
- fold DepthClipControl into PrimitiveState
- drop Descriptor from chained struct names as proposed above
- add Wgpu or Dawn at the end for implementation-specific extensions

**Comment by kainino0x** [comment]
Feedback that eliding "Descriptor" from the extension struct names is confusing:
https://chromium-review.googlesource.com/c/chromium/src/+/5691772/comment/25f7fc96_8d6b6f12/

**Comment by kainino0x** [comment]
Jul 18 meeting:

* Got feedback that new names are hard to understand  
* WGPUSurfaceFromCanvasHTMLSelectorDescriptor  
* WGPUSurfaceFromCanvasHTMLSelector  
* WGPUSurfaceDescriptorFromCanvasHTMLSelector  
* WGPUSurfaceDescriptor\_FromCanvasHTMLSelector  
  * With this in C++ we could actually namespace it?  
* KN: Try to give them descriptive names that *don't* depend on the descriptor they're attached to? (Docs tell you what they're attached to.)  
* CF: Vulkan style.  
* Brainstorm:  
*   
* WGPUCanvasHTMLSelector  
* WGPUExtCanvasHTMLSelector  
*   
* WGPUSurfaceSourceCanvasHTMLSelector  
* WGPUSurfaceSourceAndroidNativeWindow  
* WGPUSurfaceSourceMetalLayer  
*   
* **WGPUShaderSourceWGSL**  
* **WGPUShaderSourceSPIRV** \<- current caps style  
* WGPUShaderSourceSPIRV\_EXT  
* WGPUShaderSourceGLSLWgpu  
* WGPUShaderSourceGLSL\_Wgpu ?? doesn’t look right  
* **WGPUShaderSourceGLSL\_WGPU**  
* WGPUShaderSourceGLSL\_wgpu ?? Feels like typo  
* WGPUShaderSourceGLSL\_DAWN  
* WGPUShaderSourceGLSL\_GFXRS  
* WGPUShaderSourceGLSL\_RUSTYGRAPHICSMAGES  
* **WGPUShaderSourceGLSL\_Dawn**  
* WGPUShaderSourceGLSL\_NATIVE  
* WGPUShaderSourceGLSL\_STD  
* WGPUShaderSourceGLSL\_EMSCRIPTEN ??   
* WGPUShaderSourceGlslWgpu \<- proposed caps style?  
*   
* **Propose the naming conventions in bold above, get feedback**  
  * **Also need to update [\#214](https://github.com/webgpu-native/webgpu-headers/issues/214)**

(I'll expand out that proposal in a followup comment)

**Comment by kainino0x** [comment]
Rough proposal:

`WGPUPrimitiveState` extensions:
`WGPUPrimitiveDepthClipControl` or `WGPUDepthClipControl`

`WGPURenderPassDescriptor` extensions:
`WGPURenderPassMaxDrawCount` or `WGPUMaxDrawCount`

`WGPUShaderModuleDescriptor` extensions:
`WGPUShaderSourceSPIRV`
`WGPUShaderSourceWGSL`

`WGPUSurfaceDescriptor` extensions:
`WGPUSurfaceSourceAndroidNativeWindow`
`WGPUSurfaceSourceCanvasHTMLSelector_Emscripten` (see https://github.com/webgpu-native/webgpu-headers/issues/214#issuecomment-2195624598)
`WGPUSurfaceSourceMetalLayer`
`WGPUSurfaceSourceWaylandSurface`
`WGPUSurfaceSourceWindowsHWND`
`WGPUSurfaceSourceXcbWindow` or `WGPUSurfaceSourceXCBWindow`
`WGPUSurfaceSourceXlibWindow`

Other examples of how extensions could be named:
`WGPUShaderSourceGLSL_WGPU`
`WGPUShaderSourceGLSL_Dawn`

**Comment by beaufortfrancois** [comment]
> `WGPUPrimitiveState` extensions: 
> `WGPUPrimitiveDepthClipControl` or `WGPUDepthClipControl`

My understanding was that we'd fold DepthClipControl into PrimitiveState with https://github.com/webgpu-native/webgpu-headers/pull/311
@kainino0x Did I misunderstand?

**Comment by Kangz** [comment]
Yes we're folding it, but that's just for the explanation of the new naming rules.

**Comment by kainino0x** [comment]
Oh yes, I intended to note that but forgot. Edited above.

**Comment by beaufortfrancois** [comment]
I believe we can close this issue after https://github.com/webgpu-native/webgpu-headers/pull/320 gets merged.

**Comment by kainino0x** [comment]
You're right, thanks!
Dawn bug is https://crbug.com/42241174

**Comment by Kangz** [comment]
Adding the namespace in C++ is cool, but could be mildly difficult to do in Dawn, though it's not breaking for webgpu.h so it's ok to do later. That said I really don't understand why we need a `_` to separate the extension block name from the rest of the name. There are likely to be very few blocks, and they will have clear names like <vendor> or "Native". IMHO it should be a prefix to go from general (`WGPU`) to more specific. The block is the second most general thing after `WGPU`. Subjectively it makes things easy to ready while not looking weird with random underscores: 

**New enums and their enum values:** 
 - `WGPUDawnMyEnumType`
 - `WGPUDawnMyEnumType_MyValue`
 - `WGPUEnumType_DawnMyValue`
 - `wgpu::DawnEnumType`
 - `wgpu::DawnEnumType::MyValue`
 - `wgpu::EnumType::DawnMyValue`
 - Alternatively we could do the namespacing in C++ at some point `wgpu::dawn::EnumType::MyValue`.

**New methods and new objects**
 - `WGPUDawnAdapterPropertiesMemoryHeaps`
 - `wgpuDawnAdapterPropertiesMemoryHeaps_FreeMembers()`
 - `WGPUDawnMyObject`
 - `wgpuDawnMyObject_MyMethod()`
 - `wgpuObjectType_DawnMethod()`
 - `wgpu::DawnMyObject`
 - `wgpu::DawnMyObject::MyMethod()`
 - `wgpu::MyObject::DawnMethod()`
 - Likewise we could namespace in C++ `wgpu::dawn::MyObject`.

**Comment by kainino0x** [comment]
> IMHO it should be a prefix to go from general (`WGPU`) to more specific.

The Khronos-like style makes it so when something gets promoted to core, only the suffix is removed. I don't have that much preference, but we do have the problem that one of the blocks is for `wgpu` so we would have `WGPUWgpuFoo` and `wgpuWgpuBar()`

**Comment by kainino0x** [comment]
wgpu's suffix could maybe change to `wgpurs` or `gfxrs` or something.

**Comment by Kangz** [comment]
TBH `wgpuWgpu` is weird but not a big deal.

**Comment by beaufortfrancois** [comment]
Just wanted to check in and see if we've decided on a naming convention yet.

**Comment by eliemichel** [comment]
FWIW, my personal opinion: Spontaneously I would have opted for @Kangz' proposal (and agree that a `wgpuWgpu` prefix is only a minor issue, and really not the only case where having the project called "wgpu" is confusing). I see the point about the Khronos-like style, but is it used anywhere else in WebGPU, or are there any specific reasons to adopt it? How does Vulkan handle naming for extensions that add enum values, or new enums?

**Comment by kainino0x** [comment]
Nov 25 meeting:

*   almost totally missed putting this on the agenda
*   discussion: (definitely don't want double suffix or suffix stuck in the middle)
*   **OK with either, no real opinion:**
    *   `WGPUDawnMyEnumType` + `WGPUDawnMyEnumType_MyValue`
    *   `WGPUMyEnumType_Dawn` + `WGPUMyEnumType_MyValue_Dawn`

Dawn hasn't implemented suffixes yet, and I checked and wgpu-native has not yet either.
(Dawn naming is ad hoc; wgpu-native has its extensions in a separate header.)

So @Kangz since you prefer `WGPUDawnMyEnumType_MyValue` we may as well go with that (wgpu-native will probably use `WGPUWgpuMyEnumType_MyValue`)

---

> I see the point about the Khronos-like style, but is it used anywhere else in WebGPU, or are there any specific reasons to adopt it?

The only specific reason I see to adopt it is that the un-suffixed version of the name (if something becomes core) is exactly the same except for the very end. But that's not significantly different from removing the prefix between WGPU and the rest of the name. Otherwise just because it is familiar to GL/Vulkan developers.

> How does Vulkan handle naming for extensions that add enum values, or new enums?

I was pretty sure about this but I should have checked :) Vulkan does this when adding a new enum:
`VkPerformanceCounterUnitKHR`
`VK_PERFORMANCE_COUNTER_UNIT_GENERIC_KHR`

I suppose we could have inferred we wanted to use that style from the fact that we said "vulkan style", but we came to a different conclusion anyway, so no problem.

**Comment by kainino0x** [comment]
Updated #214 accordingly

**Comment by Kangz** [comment]
Yeah I think `wgpuBufferDawnClearWithDouble` (or w/e) is pretty clear that it's on `WGPUBuffer` and a Dawn method name. The examples provided above don't use any concrete name which makes it harder to separate the parts, but with `wgpuNounDawnVerb` it's pretty clear what's happening.

**Comment by kainino0x** [comment]
Ack.

**Comment by kainino0x** [comment]
To go through everything, for the record, does this seem right?

```
wgpuPrefixNewFunction
WGPUPrefixNewObject
wgpuPrefixNewObjectNewMethod
wgpuOldObjectPrefixNewMethod
WGPUPrefixNewEnum
WGPUPrefixNewEnum_OldValue
WGPUOldEnum_PrefixNewValue
WGPUPrefixNewStruct
WGPUPrefixNewBitflagType
WGPUPrefixNewBitflagType_NewValue
WGPUOldBitflagType_PrefixNewValue
WGPUPrefixNewCallback
WGPU_PREFIX_NEW_CONSTANT
```

**Comment by kainino0x** [comment]
I totally forgot the code generator actually implements the suffixing, so we need to fix that. (It's not used in this repo so it doesn't affect the core `webgpu.h`.)
